pt_miniscreen/app.py

```diff
@@ -88,59 +98,63 @@ def setup_events(self) -> None:
         )
         subscribe(AppEvent.STOP_SCREENSAVER, lambda: self.pop_tile_group())
 
-        def set_is_user_controlled(user_has_control) -> None:
-            if self.user_has_control and not user_has_control:
-                self.user_gave_back_control_event.set()
+    def set_is_user_controlled(self, user_has_control) -> None:
+        if self.user_has_control and not user_has_control:
+            self.user_gave_back_control_event.set()
+
+        logger.info(
+            f"User has {'taken' if user_has_control else 'given back'} control of the miniscreen"
+        )
+
+    def on_button_released(self):
+        if not self.state_manager.buttons_should_be_handled():
+            logger.info("State manager says that buttons should not be handled")
+            return
+
+        self.state_manager.wake()
+
+    def on_up_button_released(self):
+        logger.debug(f"Handling up button for tile group {self.current_tile_group}")
+        self.on_button_released()
 
-            logger.info(
-                f"User has {'taken' if user_has_control else 'given back'} control of the miniscreen"
-            )
+        current_tile = self.current_tile_group.current_menu_tile
+        if isinstance(current_tile, PaginatedTile):
+            current_tile.scroll_to_previous()
```

Comment:
I think it was better before, for the tile groups to implement their own 'handle_up_btn' etc, rather than for the app to know about all the tile groups and which method they want called for which button

Reply:
This breaks encapsulation: The menu tile groups shouldn't have an inherent knowledge of that kind of global behaviour, image we want to  scroll to previous due to some event other than a button click. The app is the only place we can effectively reason about when a tile group should do something.

Reply:
Well, I do agree that the app should handle the buttons and decide what to do with them, but I don't think it breaks encapsulation for it to say "oh i don't have anything to do with this up button... the current page(ythingy) should handle it".

And sure, the app might reach into the page and call `scroll_to_previous` sometimes, but the menus could also do that themselves in response to the up button.

I dunno, I can see why you want a scrolling class to have 'scroll up' methods not 'up button' methods... but really should the app handle all the dynamic behaviour? If it should then why do we have a 'handle_cancel_btn' but not up?

Reply:
I mean, it's giving responsibility of app level behaviours to the TileGroup. There is a lot of missing information when you look at the TileGroup in isolation, ignoring knowledge of how the app behaves. For instance when does the method get called? It's not every time the button is pressed, but you couldn't possibly know that from that level.

I didn't rename handle_cancel_btn because I didn't want to keep increasing the scope of the PR and refactoring every TileGroup.

Reply:
Right, but I feel this is giving responsibility of tile group behaviours to the app. Handling button presses should be behaviour of the app, but deciding to scroll should be behaviour of the tile group. So the app may handle a button press by doing it's own thing, or by passing it to the current tile group. And in turn the tile group also may directly respond to the button, or pass it through to the a page.

The app shouldn't need to know about all the possible tile groups and how they'll want to handle a particular button.

Reply:
How would the tile group know when to scroll?

Reply:
After discussion we've decided it would be simpler for all the button handling logic to be here, rather than having handlers for particular buttons on the tilegroups/tiles/pages as before. However, a more through refactor should be completed for that (eg possibly removing `handle_cancel_btn`, 'handled' checks, MenuTile class), so for now we might leave this area as it was.